### ISSUE TITLE
Migrate to C++

### ISSUE THREAD

**Issue by KOLANICH** [hashcat/hashcat#511]
I see you refactoring the code, ... so consider replacing c code with c++ one. I mean there are some places where OOP approach with virtual fuctions and abstract classes can simplify things.

**Comment by matrix** [comment]
I thought the same thing yesterday, but this is not a migration, it should be re-engineered.

**Comment by jsteube** [comment]
what do you think what i'm doing since thee weeks, lol

**Comment by matrix** [comment]
ahahahahahahahahaha :)

**Comment by matrix** [comment]
@jsteube, let me in :D

**Comment by jsteube** [comment]
Actually, I can see how C++ can be useful in some cases, but I don't see it for hashcat atm. @KOLANICH can you name some of the advantages?

What I can see when it comes to OOP approach is that the code usually results in a more ordered structure, but we can do that for C, too. That's what I meant with "what i'm doing since thee weeks". If you take a look at the commits from the last weeks you will see it.

**Comment by KOLANICH** [comment]
The advantages:
- capabitity design patterns, for example it can be possible to encapsulate cryptoalgorithms into classes and deal with them in a more uniform way (you can also do it in C putting the base object as the first field and casting the pointer)
- c++ idioms like raii
- c++ cross-platform standard library for dealing with threads, time and randomness
- more strict type checking, elimination of macroses in favour of templates
- computing tables of coefficients in compile time without storing them as they are in the source code.

The disadvantages:
- imcompatibility with C
- no standardized ABI for object oriented code which means  one doesn't simply link OO code from the libraries built with another compiler. Not a very big issue.
- the standard library tools lack some features used like setting affinity

Since C++ includes frequently used subset of C, I see no reason not to make some minor changes and start to use C++ right now.

**Comment by mosreq** [comment]
C is simpler and easier to read.
"Entities must not be multiplied beyond necessity"

**Comment by KOLANICH** [comment]
> C is simpler and easier to read.

... when used for simple hello-world-level programs.

> "Entities must not be multiplied beyond necessity"

![a popular pic from runet](https://cloud.githubusercontent.com/assets/240344/18882184/3cde489a-84ef-11e6-8933-e579be2b67c1.jpg)

> -Don't multiply entities, b1tch! Don't multiply!
> -OK, Occam, just move the razor away!

In this case it seems to be necessary because hc is much more complex than helloworld.

**Comment by matrix** [comment]
lol

**Comment by mosreq** [comment]
> ... when used for simple hello-world-level programs.

yeah? tell that to windows and linux kernel developers

**Comment by KOLANICH** [comment]
hashcat is not linux kernel and Linuse's statement was about c++03.

**Comment by jsteube** [comment]
@KOLANICH I understand the use for C++ in some projects, but why is it good for hashcat? What I meant is, can you make an example, but inside hashcat, where C++ would help us alot?

**Comment by KOLANICH** [comment]
> it can be possible to encapsulate cryptoalgorithms into classes and deal with them in a more uniform way

**Comment by polarathene** [comment]
If you're open to refactoring to another language, Rust might be another good alternative over C++. Full disclaimer, I know very little C/C++/Rust and likely a perfect example of [RIIR](http://transitiontech.ca/random/RIIR)...

Benefits in comparison to C/C++ from [this answer](https://www.quora.com/What-are-the-advantages-of-Rust-compared-to-C-C++):
- a type system feature that helps eliminate memory leaks,
- proper interfaces, called 'traits',
- better type inference,
- better support for concurrency,
- (almost) first-class functions that can be passed as arguments.

[A devs experience with C++ and Rust](http://cantrip.org/rust-vs-c++.html)

Another dev [writes about pros/cons of Rust over C++](https://plus.google.com/+nialldouglas/posts/AXFJRSM8u2t)

[Rust FAQ](https://www.rust-lang.org/en-US/faq.html)

---

Just saw the issue while checking for another and felt like mentioning the alternative, I just use hashcat so whatever you like to write it in is great :)

**Comment by KOLANICH** [comment]
Rust can be good, but it is a bit too complicated and hard-to-use. C++ is compatible with c which makes transition easy. So I think it'd be better migrate it to C++ first, reengineer, and then rewrite into rust if it worth it.

**Comment by jsteube** [comment]
I still find no good reason why to switch to C++ yet.

**Comment by KOLANICH** [comment]
I still find no good reason to stay on an old language without built-in virtual functions, inheritance and backward compatible with the old one in some large subset.

**Comment by neheb** [comment]
That sounds like a matter of opinion. Let me rephrase Jens' question.

What problem exists with hashcat that C++ solves?

**Comment by KOLANICH** [comment]
No problem specific to hc. All the apps in C I saw have the same problems: the code is bloated with pointers (solved by the convention that ```this``` is the default pointer and inheritance, allowing to call access the inherited objects without ```this```), macros (often can be solved by templates usage), resources allocation/freeing (raii) and copypasting the code (solved by inheritance and virtual functions). C++ is mainly just a syntax sugar for the things which can be (and often are) done manually in C , a bit more strict type checking and the c++ standard library.
D can be even better, but it it doesn't have enough good tools and features. Python also can be used (and is used in pyrit) since the main work is done in opencl code, but there is a drawback that you need a python interpreter/jit-compiler. It is nearly always present on linux desktop systems, but nearly never in windows ones.

**Comment by jsteube** [comment]
Closing this issue because I see no real advantage in porting to any other language. Maybe because OpenCL is so close to C it feels just natural to stick with C in the host program, too.

**Comment by KOLANICH** [comment]
OK
